# Text Classification, Lab 2: Building a model with K-Train

⚡️ Save a copy of this notebook in your Google Drive before continuing

## About this lab

The first part of this lab is identical to Lab 1, and you should be able to step through it easily.

We continue now with the goal of building an inference model for predicting whether or not a document is about "healthy living." We will use the K-Train library to do this.

## About the final project


Recall that you are working toward a final project. After completing this lab, you will want to go the extra mile and explore ways to tweak and improve your model. See the final project description for further details on what is expected.

## Imports

We're going to be using Google's Tensorflow package: 
https://www.tensorflow.org/tutorials

We're using an API wrapper for Tensorflow called ktrain. It's absolutely fabulous because it really abstracts the whole deep learning process into a workflow so easy, even a computational social scientist can do it:
https://github.com/amaiya/ktrain

In [1]:
import os
try:
  import ktrain
except:
  !pip install ktrain
  os.kill(os.getpid(), 9)
import ktrain
import pandas as pd
import numpy as np

## Mount Google Drive

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

## Set your google colab runtime to use GPU, a must for deep learning!

Runtime > Change Runtime Type > GPU

The following code snippet will show you GPU information for your runtime.

In [2]:
gpu_info = !nvidia-smi
gpu_info = '\n'.join(gpu_info)
if gpu_info.find('failed') >= 0:
  print('Not connected to a GPU')
else:
  print(gpu_info)

Mon May  2 18:26:20 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 512.15       Driver Version: 512.15       CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:05:00.0  On |                  N/A |
| 27%   40C    P8    10W / 120W |   2497MiB /  6144MiB |      1%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

## Load the data

The data file should be in your Google Drive from Lab 1.

In [3]:
reviews = pd.read_json("news_category_trainingdata.json")

## Inspect the data

In [4]:
reviews.head()

category                                           headline  \
0          CRIME  There Were 2 Mass Shootings In Texas Last Week...   
1  ENTERTAINMENT  Will Smith Joins Diplo And Nicky Jam For The 2...   
2  ENTERTAINMENT    Hugh Grant Marries For The First Time At Age 57   
3  ENTERTAINMENT  Jim Carrey Blasts 'Castrato' Adam Schiff And D...   
4  ENTERTAINMENT  Julianna Margulies Uses Donald Trump Poop Bags...   

           authors                                               link  \
0  Melissa Jeltsen  https://www.huffingtonpost.com/entry/texas-ama...   
1    Andy McDonald  https://www.huffingtonpost.com/entry/will-smit...   
2       Ron Dicker  https://www.huffingtonpost.com/entry/hugh-gran...   
3       Ron Dicker  https://www.huffingtonpost.com/entry/jim-carre...   
4       Ron Dicker  https://www.huffingtonpost.com/entry/julianna-...   

                                   short_description       date  
0  She left her husband. He killed their children... 2018-05-26  
1                           Of course it has a song. 2018-05-26  
2  The actor and his longtime girlfriend Anna Ebe... 2018-05-26  
3  The actor gives Dems an ass-kicking for not fi... 2018-05-26  
4  The "Dietland" actress said using the bags is ... 2018-05-26

## Prepare the data

Most machine learning tools in Python accept one field/column/string. So we have to merge our two text column. Let's separate it with a space.

In [5]:
reviews['combined_text'] = reviews['headline'] + ' ' + reviews['short_description']

The first thing we need to do is prepare the data. Specifically, we have a categorical column that we want to turn into a "is this article healthy living?" column. That is, when an article is about healthy living, it should have a 1, when it's anything else, it should be a 0.

In [37]:
reviews[reviews['category'].str.contains("HEALTHY LIVING")]

category                                           headline  \
7578    HEALTHY LIVING  To The People Who Say ‘I’m Tired’ When Someone...   
7693    HEALTHY LIVING  Eating Shake Shack Made Me Feel Healthier Than...   
7747    HEALTHY LIVING  How To Stay Updated On The News Without Losing...   
7927    HEALTHY LIVING  27 Perfect Tweets About Whole30 That Will Make...   
7934    HEALTHY LIVING         The Real Reason Your Hands Are Always Cold   
...                ...                                                ...   
124913  HEALTHY LIVING  Why You Need Both a 'Bouncer' and a 'Bartender...   
124914  HEALTHY LIVING  How Video Games Can Improve Dialogue on Mental...   
124925  HEALTHY LIVING  Wake-Up Calls Inspired My Change From Overdriv...   
124950  HEALTHY LIVING        Loving a Narcissist Without Losing Yourself   
124988  HEALTHY LIVING                            Reasons Not to Be Happy   

                                                  authors  \
7578    The Mighty, ContributorWe face disability, dis...   
7693    Colleen Werner, ContributorCampus Editor-at-Large   
7747                                       Lindsay Holmes   
7927                                       Lindsay Holmes   
7934    Refinery29, ContributorThe #1 new-media brand ...   
...                                                   ...   
124913  Elizabeth Grace Saunders, ContributorFounder, ...   
124914         Mona Shattell, Contributornurse researcher   
124925  Jane Shure, ContributorLeadership Coach, Psych...   
124950  Nancy Colier, ContributorPsychotherapist, inte...   
124988  Mindy Utay, Contributor"Calming Life's Conflicts"   

                                                     link  \
7578    https://www.huffingtonpost.com/entry/to-the-pe...   
7693    https://www.huffingtonpost.com/entry/eating-sh...   
7747    https://www.huffingtonpost.com/entry/anxiety-f...   
7927    https://www.huffingtonpost.com/entry/tweets-ab...   
7934    https://www.huffingtonpost.com/entry/the-real-...   
...                                                   ...   
124913  https://www.huffingtonpost.com/entry/happy-hea...   
124914  https://www.huffingtonpost.com/entry/mental-il...   
124925  https://www.huffingtonpost.com/entry/wake-up-c...   
124950  https://www.huffingtonpost.com/entry/narcissis...   
124988  https://www.huffingtonpost.com/entry/happiness...   

                                        short_description       date  \
7578    When you feel like this, it’s important to kno... 2018-01-16   
7693    I can vividly remember the first time I felt f... 2018-01-12   
7747       Because it's only becoming more of a struggle. 2018-01-12   
7927    "The only Whole30 I want to participate in is ... 2018-01-10   
7934    Essentially, your hands are kept warm thanks t... 2018-01-10   
...                                                   ...        ...   
124913  Instead of judging whether you made the right ... 2014-04-18   
124914  While there are strong arguments for the games... 2014-04-18   
124925  My wake-up call marching orders were clear: No... 2014-04-18   
124950  It is very difficult for some people to see an... 2014-04-18   
124988  Our thoughts and feelings are powerful, but ma... 2014-04-18   

                                            combined_text  healthy  
7578    To The People Who Say ‘I’m Tired’ When Someone...        1  
7693    Eating Shake Shack Made Me Feel Healthier Than...        1  
7747    How To Stay Updated On The News Without Losing...        1  
7927    27 Perfect Tweets About Whole30 That Will Make...        1  
7934    The Real Reason Your Hands Are Always Cold Ess...        1  
...                                                   ...      ...  
124913  Why You Need Both a 'Bouncer' and a 'Bartender...        1  
124914  How Video Games Can Improve Dialogue on Mental...        1  
124925  Wake-Up Calls Inspired My Change From Overdriv...        1  
124950  Loving a Narcissist Without Losing Yourself It...        1  

In [38]:
reviews[reviews['category'].str.contains("WELLNESS")]

category                                           headline  \
124989  WELLNESS                Why Overeating Doesn't Make You Fat   
124990  WELLNESS  14 Habits Of People With A Healthy Relationshi...   
124993  WELLNESS           5 Things That Could Be Stealing Your Joy   
124994  WELLNESS                                Moments Make a Life   
124996  WELLNESS                                          Fat Facts   
...          ...                                                ...   
200797  WELLNESS  Shoveling Snow? How to Protect Your Back (And ...   
200799  WELLNESS  7 Reasons Working Too Much Is Bad For Your Health   
200800  WELLNESS   The Sleep Library: 11 Soothing Books For Bedtime   
200802  WELLNESS                   The Benefits of Caring for a Pet   
200805  WELLNESS  This Is Only the Beginning: Surprising Advice ...   

                                                  authors  \
124989  Mark Hyman, MD, Contributor\nPracticing Physician   
124990                                        Sarah Klein   
124993  Amanda Gore, Contributor\nInternational keynot...   
124994  Megan McDonough, Contributor\nCEO, Wholebeing ...   
124996  Maria Rodale, Contributor\nCEO and Chairman of...   
...                                                   ...   
200797                                                      
200799                                        Sarah Klein   
200800                                                      
200802  Rita Altman, R.N., Contributor\nSenior Vice Pr...   
200805  Ellie Knaus, Contributor\nAtomic Moms Podcast ...   

                                                     link  \
124989  https://www.huffingtonpost.com/entry/overeatin...   
124990  https://www.huffingtonpost.com/entry/healthy-r...   
124993  https://www.huffingtonpost.com/entry/joy_us_5b...   
124994  https://www.huffingtonpost.com/entry/moments-m...   
124996  https://www.huffingtonpost.com/entry/fat-facts...   
...                                                   ...   
200797  https://www.huffingtonpost.com/entry/shovel-sn...   
200799  https://www.huffingtonpost.com/entry/overtime-...   
200800  https://www.huffingtonpost.comhttp://www.oprah...   
200802  https://www.huffingtonpost.com/entry/pets-seni...   
200805  https://www.huffingtonpost.com/entry/life-tips...   

                                        short_description       date  \
124989  If you take the right approach to ending your ... 2014-04-17   
124990  After a long list of rules and habits like the... 2014-04-17   
124993  Listen to yourself all day today -- are you an... 2014-04-17   
124994  Consciously constructing positive habits -- or... 2014-04-17   
124996  Human fat cells, photo by Dr. Peeke Here's how... 2014-04-17   
...                                                   ...        ...   
200797  Work up to it. Warm up inside, and incorporate... 2012-01-28   
200799  If you're the last one to leave the office, or... 2012-01-28   
200800  Do you toss and turn until you finally sit up ... 2012-01-28   
200802  For the young as well as the old, especially i... 2012-01-28   
200805  My great-aunt Ida loves to say, "This is only ... 2012-01-28   

                                            combined_text  healthy  
124989  Why Overeating Doesn't Make You Fat If you tak...        0  
124990  14 Habits Of People With A Healthy Relationshi...        0  
124993  5 Things That Could Be Stealing Your Joy Liste...        0  
124994  Moments Make a Life Consciously constructing p...        0  
124996  Fat Facts Human fat cells, photo by Dr. Peeke ...        0  
...                                                   ...      ...  
200797  Shoveling Snow? How to Protect Your Back (And ...        0  
200799  7 Reasons Working Too Much Is Bad For Your Hea...        0  
200800  The Sleep Library: 11 Soothing Books For Bedti...        0  
200802  The Benefits of Caring for a Pet For the young...        0  
200805  This Is Only the Beginning: Surprising Advice ...        

In [97]:
#reviews['health_Well'] = np.where((reviews['category'] == 'HEALTHY LIVING'), 1, 0)
reviews['health_Well'] = np.where((reviews['category'] == 'HEALTHY LIVING') | (reviews['category'] == 'WELLNESS'), 1, 0)

In [103]:
reviews.health_Well.describe()

count    200853.000000
mean          0.122084
std           0.327384
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max           1.000000
Name: health_Well, dtype: float64

In [105]:
reviews.health_Well[reviews.health_Well == 1].describe()

count    24521.0
mean         1.0
std          0.0
min          1.0
25%          1.0
50%          1.0
75%          1.0
max          1.0
Name: health_Well, dtype: float64

So we have 24521 artikels about Healthy living or Wellness. This matches the data description

## Balance the data

To create a balanced data set that includes all of the health living articles, set sample_amount to the total number of those articles.

In Lab 1, you balanced the data for the full set of healthy living articles. In the interest of getting through Lab 2 more quickly (in terms of training time for the model), we will use a smaller sample, of just 1000 articles per class. After completing the lab, consider increasing the sample size to see if you can get improvements on the model performance. Of course, be prepared for longer training times when you do that.

In [106]:
# We have replaced the sample count with a smaller number in order to expedite
# the completion of the lab. For your final project, you will want to use the
# full balanced document set which is determined by this commented line:
#sample_amount =  len(reviews[reviews["healthy"] == 1]) # the total number of healthy living articles

sample_amount = 10000

healthy = reviews[reviews['health_Well'] == 1].sample(n=sample_amount)
not_healthy = reviews[reviews['health_Well'] == 0].sample(n=sample_amount)

In [107]:
review_sample = pd.concat([healthy,not_healthy])

In [108]:
review_sample.describe()

healthy   health_Well
count  20000.000000  20000.000000
mean       0.136650      0.500000
std        0.343486      0.500013
min        0.000000      0.000000
25%        0.000000      0.000000
50%        0.000000      0.500000
75%        0.000000      1.000000
max        1.000000      1.000000

# On to Lab 2: Test, Tune and Save Models

Here, you will tune and train a predictor model for classifying healthy-living articles. After completing this lab, complete the Lab Quiz by entering your precision and recall values from the validation report for both the negative and positive classes.

In [109]:
target_names = ['NOT health_Well','health_Well']

---

### Experimenting with different transformers

For purposes of this lab, we are using the **distilbert-base-uncased** transformer model. Other models you might try for your final project include:

 * roberta-base
 * bert-base-uncased
 * distilroberta-base

See all the models here: https://huggingface.co/transformers/pretrained_models.html

Some work, some dont, try at your own risk.

---

In [110]:
train, val, preprocess = ktrain.text.texts_from_df(
    review_sample,
    "combined_text",
    label_columns=["health_Well"],
    val_df=None,
    max_features=20000,
    maxlen=512,
    val_pct=0.1,
    ngram_range=1,
    preprocess_mode="distilbert",
    verbose=1
)

['not_health_Well', 'health_Well']
        not_health_Well  health_Well
117099              1.0          0.0
88591               1.0          0.0
32100               0.0          1.0
36362               1.0          0.0
20359               0.0          1.0
['not_health_Well', 'health_Well']
        not_health_Well  health_Well
65180               1.0          0.0
94678               1.0          0.0
147468              1.0          0.0
168047              0.0          1.0
177528              0.0          1.0
preprocessing train...
language: en
train sequence lengths:
	mean : 32
	95percentile : 58
	99percentile : 68


Is Multi-Label? False
preprocessing test...
language: en
test sequence lengths:
	mean : 32
	95percentile : 58
	99percentile : 67


In [111]:
model = preprocess.get_classifier()
learner = ktrain.get_learner(model, train_data=train, val_data=val, batch_size=4)

In [ ]:
learner.lr_find(max_epochs=6)

simulating training for different learning rates... this may take a few moments...
Epoch 1/6
 791/4500 [====>.........................] - ETA: 29:39 - loss: 0.6906 - accuracy: 0.5275

In [ ]:
learner.lr_plot()

Now, use the tuned learner to train the best model.

Here, we define a limit of 10 epochs, but in reality, this should stop much sooner due to early stopping.

In [ ]:
history=learner.autofit(
    1e-4,
    checkpoint_folder='checkpoint',
    epochs=10,
    early_stopping=True
)

Get the predictor

In [ ]:
predictor = ktrain.get_predictor(learner.model, preproc=preprocess)

Optionally, uncomment this code to save the predictor and reload it later. Note, the saved models can be quite large and may quickly use up space on your Google Drive.

In [26]:
#predictor.save("drive/MyDrive/MSDSTextClassification_Lab2.healthy_living")

In [ ]:
validation = learner.validate(val_data=val, print_report=True)

---

## 🧐 Lab Quiz Questions 1-4

Enter the following values from the validation report into the Lab Quiz:

 1. precision for non healthy-living articles
 2. recall for non healthy-living articles
 3. precision for healthy-living articles
 4. recall for healthy-living articles

 ---


Keep in mind that we've reduced the training set for the sake of expediency. For your final analysis and project, you should complete a run of the full data set. Pay attention to the impact of the input data on the performance of the final model (i.e. the validation scores)

# Inspecting the drivers of prediction

No matter what the supervised machine learning model, you always want to peak under the hood to see what features are driving prediction. That is, what words sway the outcome of the prediction. It's harder to inspect a neural network. Because all of the layers of a neural network aren't really interpretable to the human eye. 

Currently, the best practice I've found is a little tool Explainable AI:
https://alvinntnu.github.io/python-notes/nlp/ktrain-tutorial-explaining-predictions.html

In [ ]:
!pip3 install -q git+https://github.com/amaiya/eli5@tfkeras_0_10_1

Let's go ahead and make a little set of test documents to check out

In [ ]:
test_docs = [
'Stress May Be Your Heart’s Worst Enemy Psychological stress activates the fear center in the brain, setting into motion a cascade of reactions that can lead to heart attacks and strokes.',
'Exercising to Slim Down? Try Getting Bigger. It’s high time for women to reclaim the real strength behind exercise.',
'What Are Your Food Resolutions for the New Year? Join us for the Eat Well Challenge starting in January.',
'Why We All Need to Have More Fun. Prioritizing fun may feel impossible right now. But this four-step plan will help you rediscover how to feel more alive.',
'Cuomo Will Not Be Prosecuted in Groping Case, Albany D.A. Says. The district attorney described the woman who said former Gov. Andrew Cuomo had groped her as “credible,” but added that proving her allegation would be difficult.',
'A Film Captures Jewish Life in a Polish Town Before the Nazis Arrived. A documentary based on a home movie shot by an American in 1938 provides a look at the vibrancy of a Jewish community in Europe just before the Holocaust.' 
             ]

In [ ]:
for i, text in enumerate(test_docs):
  probs = predictor.predict(text, return_proba=True)
  print("---------------------------")
  print('The probability this is healthy is %s' % probs[1])
  print(text)

*These* are pretty obvious examples, but it works exactly as expected!

In [ ]:
predictor.explain('Diversity is the key to a healthy society. Here is what we need to do to make america a more equitable place to live for all.')

But you can see, this algorithm is far from perfect. Here you can see that it's probably got too high of an emphasis on the word "healthy."

So what would I do next? Well, given that it's over reacting to worrds like health and equitable, I'd try introducing more negative examples into the data, times where healthy is used outside of health and wellness news. We can do this by changing our sample from 50/50 to something like 20/80, but of course, the more documents we process, the longer this model is going to take to make!